### ISSUE TITLE
Partial indexes doesn't work well with prepared statements

### ISSUE THREAD

**Issue by ArsileLuci** [cockroachdb/cockroach#66465]
**Describe the problem**

When partial index hinted to be used within prepared `UPDATE` statement query refuses to be executed with error: `<index_name> is a partial index that does not contain all the rows needed to execute this query` while the same query without preparation works well.

**To Reproduce**

*1. Create table with partial index*

```sql
CREATE TABLE partial_index_test (
  id      UUID NOT NULL PRIMARY KEY,
  meta    JSONB NOT NULL default '{}',
  flag    BOOL NOT NULL,
  at      TIMESTAMPTZ NOT NULL
);

CREATE INDEX my_partial_idx ON partial_index_test (at) WHERE (meta -> 'test') = '1'
```

*2. Try to run query without preparation*
```sql
UPDATE partial_index_test@my_partial_idx
SET meta = jsonb_set(
  meta,
  ARRAY['partial_index'],
  '"working"'::JSONB
)
WHERE (meta -> 'test') = '1'
AND at >= now();
```
```sql
UPDATE partial_index_test@my_partial_idx
SET flag = true
WHERE (meta -> 'test') = '1'
AND at >= now();
```
```sql
SELECT id FROM partial_index_test@my_partial_idx
WHERE (meta -> 'test') = '1'
AND at <= now();
```

And all of them just work.

*3. But if this queries are prepared*
```sql
PREPARE prep1 AS UPDATE partial_index_test@my_partial_idx
SET meta = jsonb_set(
  meta,
  ARRAY['partial_index'],
  '"working"'::JSONB
)
WHERE (meta -> 'test') = '1'
AND at >= now();

EXECUTE prep1
```

```sql
PREPARE prep2 AS UPDATE partial_index_test@my_partial_idx
SET flag = true
WHERE (meta -> 'test') = '1'
AND at >= now();

EXECUTE prep2
```

```sql
PREPARE prep3 AS SELECT id FROM partial_index_test@my_partial_idx
WHERE (meta -> 'test') = '1'
AND at <= now();

EXECUTE prep3;
```

Error: `my_partial_idx is a partial index that does not contain all the rows needed to execute this query` is thrown.

**Expected behavior**

Prepared and not prepared statements behave identically.

**Environment:**
 - CockroachDB version: 21.1.1

**Additional context**
`SELECT` also partially affected by this issue but in some cases it just works, I'll extend this issue with `SELECT` example after find one. 

P.S. Extended with `SELECT` example, and tested more statements, prepared statements working well with hinted partial index while it contains only constant values. For example this works fine:
```sql
PREPARE prep9 AS 
SELECT id FROM partial_index_test@my_partial_idx
WHERE (meta -> 'test') = '1'
AND at::bigint <= 100000;

EXECUTE prep9;
```

**Comment by rafiss** [comment]
@ArsileLuci Thanks for reporting this! Not sure where the issue is yet, but since partial indexes is a new feature, we will look into this.

**Comment by mgartner** [comment]
@ArsileLuci Thanks again for reporting this issue! We've identified and fixed the issue. I expect the fix to be included in v21.1.4.
